axum/src/extract/state.rs

```diff
@@ -51,6 +51,26 @@ use std::{
 ///
 /// [order-of-extractors]: crate::extract#the-order-of-extractors
 ///
+/// Note that `State` should be provided last when building a [`Router`]. This means this code won't compile:
+///
+/// ```compile_fail
+/// # use axum::{Router, routing::get, extract::State};
+/// # #[derive(Clone)]
+/// # struct AppState {}
+/// #
+/// # let state = AppState {};
+/// #
+/// let app = Router::new()
+///     // Dont do this! Place `with_state` after `route`, `nest` or `merge`.
+///     .with_state(state)
+///     .route("/", get(handler));
+/// #
+/// # async fn handler(State(state): State<AppState>) {}
+///
+/// // Compilation will fail here
+/// let service = app.into_make_service();
+/// ```
+///
```

Comment:
I think we should move this example to the docs for `Router::with_state` since that's the method that's important in the example.

Reply:
I think it's easier to discover on that page since that's what's linked from the docs index, not the with_state method. Do you insist on this? If so i will move it, and simply place a shorter sentence on the State docs if that's alright with you.

Reply:
Yes I think it should be moved.